explorer/interpreter/type_checker.cpp

```diff
@@ -2920,6 +2920,372 @@ auto TypeChecker::TypeCheckPattern(
+      if (auto* tuple = dyn_cast<TupleValue>(value)) {
+        auto* tuple_type = cast<TupleValue>(type_);
+        assert(tuple->elements().size() == tuple_type->elements().size());
+        for (size_t i = 0; i != tuple->elements().size(); ++i) {
+          out.push_back(
+              AbstractPattern(tuple->elements()[i], tuple_type->elements()[i]));
+        }
+      } else if (auto* alt = dyn_cast<AlternativeValue>(value)) {
+        out.push_back(AbstractPattern(
+            &alt->argument(),
+            *cast<ChoiceType>(type_)->FindAlternative(alt->alt_name())));
+      }
+    }
+  }
+
+  auto value() const -> const Value& { return *value_.get<const Value*>(); }
+
+  auto type() const -> const Value& {
+    assert(kind() == Compound);
+    return *type_;
+  }
+
+ private:
+  struct alignas(8) WildcardTag {};
+  AbstractPattern(WildcardTag)
+      : value_(static_cast<const WildcardTag*>(nullptr)), type_(nullptr) {}
+
+  void Set(Nonnull<const Pattern*> pattern) {
+    type_ = &pattern->static_type();
+    switch (pattern->kind()) {
+      case PatternKind::AddrPattern:
+      case PatternKind::AutoPattern:
+      case PatternKind::BindingPattern:
+      case PatternKind::GenericBinding:
+        value_ = static_cast<const WildcardTag*>(nullptr);
+        break;
+
+      case PatternKind::TuplePattern:
+      case PatternKind::AlternativePattern:
+        value_ = pattern;
+        break;
+
+      case PatternKind::ExpressionPattern:
+        value_ = &pattern->value();
+        break;
+
+      case PatternKind::VarPattern:
+        Set(&cast<VarPattern>(pattern)->pattern());
+        break;
+    }
+  }
+
+ private:
+  // The underlying pattern: either a syntactic pattern, or a constant value,
+  // or a placeholder indicating that this is a wildcard pattern.
+  llvm::PointerUnion<Nonnull<const Pattern*>, Nonnull<const Value*>,
+                     const WildcardTag*>
+      value_;
+  // Values don't always know their types, so store the type here. We only
+  // really need it for the `const Value*` case but also store it for the
+  // `const Pattern*` case for convenience.
+  const Value* type_;
+};
+
+// A matrix of patterns, used for determining exhaustiveness and redundancy of
+// patterns in a match statement.
+//
+// See http://moscova.inria.fr/~maranget/papers/warn/index.html for details on
+// the algorithm used here.
+class PatternMatrix {
+ public:
+  // Add a pattern vector row to this collection of pattern vectors.
+  void Add(std::vector<AbstractPattern> pattern_vector) {
+    assert(matrix_.empty() || matrix_[0].size() == pattern_vector.size());
+    matrix_.push_back(std::move(pattern_vector));
+  }
+
+  // The maximum number of times we will consider all alternatives when
+  // recursively expanding the pattern. Allowing this to happen an arbitrary
+  // number of times leads to exponential growth in the runtime of the
+  // algorithm.
+  static constexpr int MaxExponentialDepth = 8;
```

Comment:
Note, this should be at the top of the class declaration order because it's a static constant.

Reply:
Done.